packages/web/components/related-assets/related-assets.tsx

```diff
@@ -0,0 +1,170 @@
+import { RatePretty } from "@keplr-wallet/unit";
+import { ObservableQueryPool } from "@osmosis-labs/stores";
+import { observer } from "mobx-react-lite";
+import Image from "next/image";
+import Link from "next/link";
+import { FunctionComponent } from "react";
+import { useTranslation } from "react-multi-lang";
+
+import { useAssetInfoConfig } from "~/hooks";
+import { useStore } from "~/stores";
+import { CoinBalance } from "~/stores/assets";
+
+import caretDown from "../../public/icons/caret-down.svg";
+
+const numberOfAssetsToDisplay = 8;
+
+const RelatedAssets: FunctionComponent<{
+  memoedPools: ObservableQueryPool[];
+  tokenDenom: string;
+}> = observer(({ memoedPools, tokenDenom }) => {
+  const t = useTranslation();
+
+  const { assetsStore, queriesExternalStore } = useStore();
+
+  const balances = assetsStore.nativeBalances.concat(
+    assetsStore.unverifiedIbcBalances
+  );
+
+  const assetsToDisplay = new Set<string>();
+  for (const pool of memoedPools) {
+    if (assetsToDisplay.size === numberOfAssetsToDisplay) {
+      break;
+    }
+    if (pool.poolAssets.some((asset) => asset.amount.denom === tokenDenom)) {
+      assetsToDisplay.add(
+        pool.poolAssets.filter((asset) => asset.amount.denom !== tokenDenom)[0]
+          .amount.denom
+      );
+    }
+  }
```

Comment:
We do not need a memo, because we are directly computing the mobx state, the result is already memoized , but we can improve the code to move the logic and avoid having it inline

Reply:
I also did a bit of refactoring: cba0c55456f23e9c76d4be5e1a18dd0ba2960177